osu.Desktop/OsuGameDesktop.cs

```diff
@@ -146,9 +159,50 @@ protected override string LocateBasePath()
                 return null;
             }
 
-            public StableStorage(DesktopGameHost host)
-                : base(string.Empty, host)
+            /// <summary>
+            /// osu! stable Songs folder can be moved to another location by changing the 'BeatmapDirectory' setting in the stable configuration file.
+            /// This locates the Songs folder location for this stable instalation and sets flags if the songs folder location has been customized.
+            /// </summary>
+            /// <returns></returns>
+            private string locateSongsDirectory()
             {
+                //we get the user config file.
+                var configFile = GetStream(GetFiles(".", "osu!.*.cfg").First());
+                var textReader = new StreamReader(configFile);
+
+                while (!textReader.EndOfStream)
+                {
+                    var line = textReader.ReadLine();
+
+                    if (line.StartsWith("BeatmapDirectory"))
+                    {
+                        var dir = line.Split('=')[1].TrimStart();
+                        if (!dir.Equals(default_songs_path, StringComparison.Ordinal))
+                        {
+                            usesCustomSongsPath = true;
+                            return Path.GetFullPath(dir);
+                        }
+                    }
+                }
+
+                return null;
+            }
+
+            public override Storage GetStorageForDirectory(string directory)
+            {
+                if (directory.Equals(default_songs_path, StringComparison.Ordinal) && usesCustomSongsPath)
+                    return new StableSongsStorage(songsPath, host);
```

Comment:
So I suppose this allocation is the only reason that `usesCustomSongsPath` exists (to not reallocate storage in the default case) - but maybe we can just have a `Lazy<StableSongsStorage>` and return that? Not sure about thread safety of operations though, would have to check to be 100% sure.

Reply:
sounds like this needs `ThreadLocal<StableSongsStorabe>` instead?

Reply:
Why use a `Lazy` here though? This code path is reached once (when `ArchiveModelManager.ImportFromStableAsync()` is called), so I guess it should be fine to allocate directly the Storage instead of wrapping it in a Lazy.

I see your point in using a `Lazy` to cache the instance if this exact code path was to execute several times, but can we consider it really necessary  (what are the chances that the user will re-import his beatmaps two times in a row)

Unless i'm missing something?

Reply:
I was just mostly confused at why that was written in this way and figured that might have been the reason. Would have to look at how heavy the allocation cost is to say if this is worth the trouble or not.